File: src/hyperliquid_python/helpers/hyperliquid_helpers.py
import os
from hyperliquid.exchange import Exchange
from hyperliquid.info import Info
from hyperliquid.utils import constants
from helpers.error_handler import AppError
from constants.http import HTTP
from repositories.dynamo_repository import get_order

def get_env_config():
    from helpers.config_helpers import get_secret
    
    # Use environment var to determine the SSM path
    ssm_path = os.environ.get("SSM_PRIVATE_KEY_PATH", "/hl/private_key")
    private_key = get_secret("HYPERLIQUID_PRIVATE_KEY", ssm_path)
    
    user_address = os.environ.get("HYPERLIQUID_USER_ADDRESS")
    is_testnet = os.environ.get("HYPERLIQUID_TESTNET") == "true"
    
    if not private_key:
        raise ValueError(f"HYPERLIQUID_PRIVATE_KEY not configured in env or SSM ({ssm_path})")
    if not user_address:
        raise ValueError("HYPERLIQUID_USER_ADDRESS not configured")
        
    return private_key, user_address, is_testnet

def create_clients(private_key: str, is_testnet: bool):
    from eth_account import Account
    import logging
    
    # Configure logging
    logging.basicConfig(level=logging.INFO)
    
    account = Account.from_key(private_key)
    base_url = constants.TESTNET_API_URL if is_testnet else constants.MAINNET_API_URL

    # Both Info and Exchange constructors need spot_meta bypass on testnet
    # This avoids the "list index out of range" error in hyperliquid-python-sdk
    safe_spot_meta = {"universe": [], "tokens": []}
    info_client = Info(base_url, skip_ws=True, spot_meta=safe_spot_meta)
    exchange_client = Exchange(account, base_url, spot_meta=safe_spot_meta)

    return exchange_client, info_client

def extract_order_id(status: dict) -> int:
    if not status:
        raise AppError("Order status is missing", HTTP.BAD_REQUEST)
        
    resting = status.get('resting')
    if resting and resting.get('oid'):
        return resting.get('oid')
        
    filled = status.get('filled')
    if filled and filled.get('oid'):
        return filled.get('oid')
        
    raise AppError("Order ID not found in response", HTTP.BAD_REQUEST)

def get_asset_info(info_client: Info, symbol: str):
    meta = info_client.meta()
    
    asset_info = None
    asset_id = -1
    
    for idx, asset in enumerate(meta.get('universe', [])):
        if asset.get('name') == symbol:
            asset_info = asset
            asset_id = idx
            break
            
    if not asset_info:
        raise AppError(f"Asset {symbol} not found", HTTP.BAD_REQUEST)
        
    return asset_info, asset_id

def get_strategy_position(info_client: Info, user_address: str, symbol: str, strategy: str):
    user_data = info_client.user_state(user_address)
    
    position_data = None
    for pos in user_data.get('assetPositions', []):
        if pos.get('position', {}).get('coin') == symbol:
            position_data = pos
            break
            
    if not position_data:
        raise AppError(f"No open position found for {symbol}", HTTP.BAD_REQUEST)
        
    open_order = get_order(symbol=symbol, strategy=strategy, status="open")
    if not open_order:
        raise AppError(f"No open order found for {symbol}/{strategy}", HTTP.BAD_REQUEST)
        
    print("Order id", open_order.oid)
    strategy_position = open_order.quantity
    is_buy = open_order.order_type == "BUY"
    id = open_order.id
    oid = open_order.oid
    
    full_position = abs(float(position_data.get('position', {}).get('szi', '0')))
    if full_position < strategy_position:
        raise AppError(f"Position size ({full_position}) is less than order size ({strategy_position})", HTTP.BAD_REQUEST)
        
    return strategy_position, is_buy, id, oid

def get_order_pnl(info_client: Info, user_address: str, oid: str):
    try:
        user_fills = info_client.user_fills(user_address)
        
        order_trades = [fill for fill in user_fills if str(fill.get('oid')) == oid]
        if not order_trades:
            return {"totalPnl": 0, "totalFees": 0, "netPnl": 0, "trades": []}
            
        total_pnl = 0.0
        total_fees = 0.0
        
        for trade in order_trades:
            pnl = float(trade.get('closedPnl', '0'))
            fee = abs(float(trade.get('fee', '0')))
            
            total_pnl += pnl
            total_fees += fee
            
        net_pnl = total_pnl - total_fees
        
        return {
            "totalPnl": total_pnl,
            "totalFees": total_fees,
            "netPnl": net_pnl,
            "trades": order_trades
        }
    except Exception as e:
        print("[get_order_pnl] Error fetching order PnL:", e)
        raise e

File: src/hyperLiquid_ts_deprecated/helpers/hyperliquid.helpers.ts
import * as hl from "@nktkas/hyperliquid";
import { AppError } from "./errorHandler";
import { HTTP } from "../constants/http";
import { getOrder } from "../repositories/dynamo.repository";

/**
 * Get environment configuration for HyperLiquid
 * @returns Private key, user address, and testnet flag
 */
export function getEnvConfig() {
    const privateKey = process.env.HYPERLIQUID_PRIVATE_KEY;
    const userAddress = process.env.HYPERLIQUID_USER_ADDRESS;
    const isTestnet = process.env.HYPERLIQUID_TESTNET === "true";

    if (!privateKey) throw new Error("HYPERLIQUID_PRIVATE_KEY not configured");
    if (!userAddress) throw new Error("HYPERLIQUID_USER_ADDRESS not configured");

    return { privateKey, userAddress, isTestnet };
}

/**
 * Extract order ID from HyperLiquid order status response
 * @param status - Order status response from HyperLiquid API
 * @returns Order ID (OID)
 * @throws AppError if order ID cannot be extracted
 */
export function extractOrderId(status: any): number {
    if (!status) throw new AppError("Order status is missing", HTTP.BAD_REQUEST);

    if (status.resting?.oid) return status.resting.oid;
    if (status.filled?.oid) return status.filled.oid;

    throw new AppError("Order ID not found in response", HTTP.BAD_REQUEST);
}

/**
 * Create HyperLiquid API clients
 * @param privateKey - Wallet private key
 * @param isTestnet - Whether to use testnet
 * @returns Exchange and info clients
 */
export function createClients(privateKey: string, isTestnet: boolean) {
    const transport = new hl.HttpTransport({ isTestnet });
    return {
        exchangeClient: new hl.ExchangeClient({ wallet: privateKey, transport }),
        infoClient: new hl.InfoClient({ transport })
    };
}

/**
 * Get asset metadata from HyperLiquid
 * @param infoClient - HyperLiquid InfoClient
 * @param symbol - Trading symbol
 * @returns Asset info and asset ID
 */
export async function getAssetInfo(infoClient: hl.InfoClient, symbol: string) {
    const meta = await infoClient.meta();
    const assetInfo = meta.universe.find(asset => asset.name === symbol);

    if (!assetInfo) {
        throw new AppError(`Asset ${symbol} not found`, HTTP.BAD_REQUEST);
    }

    return {
        assetInfo,
        assetId: meta.universe.indexOf(assetInfo)
    };
}

/**
 * Get user's current position and validate against database order
 * @param infoClient - HyperLiquid InfoClient
 * @param userAddress - User's wallet address
 * @param symbol - Trading symbol
 * @param strategy - Strategy name
 * @returns Strategy position size, direction, and partition key
 */
export async function getStrategyPosition(
    infoClient: hl.InfoClient,
    userAddress: string,
    symbol: string,
    strategy: string
) {
    const userData = await infoClient.webData2({ user: userAddress as `0x${string}` });
    const positionData = userData.clearinghouseState.assetPositions.find(
        pos => pos.position.coin === symbol
    );

    if (!positionData) {
        throw new AppError(`No open position found for ${symbol}`, HTTP.BAD_REQUEST);
    }

    const openOrder = await getOrder({ symbol, strategy, status: "open" });
    if (!openOrder) {
        throw new AppError(`No open order found for ${symbol}/${strategy}`, HTTP.BAD_REQUEST);
    }
    console.log("Order id", openOrder.oid);
    const strategyPosition = openOrder.quantity;
    const isBuy = openOrder.order_type === "BUY";
    const id = openOrder.id;
    const oid = openOrder.oid;

    const fullPosition = Math.abs(parseFloat(positionData.position.szi));
    if (fullPosition < strategyPosition) {
        throw new AppError(
            `Position size (${fullPosition}) is less than order size (${strategyPosition})`,
            HTTP.BAD_REQUEST
        );
    }

    return { strategyPosition, isBuy, id, oid };
}

/**
 * Get PnL from an order by fetching its fills
 * @param infoClient - HyperLiquid InfoClient
 * @param userAddress - User's wallet address
 * @param oid - Order ID from HyperLiquid
 * @returns PnL and fees from the order
 */
export async function getOrderPnl(
    infoClient: hl.InfoClient,
    userAddress: string,
    oid: string
): Promise<{ totalPnl: number; totalFees: number; netPnl: number; trades: any[] }> {
    try {
        const userFills = await infoClient.userFills({ user: userAddress as `0x${string}` });

        const orderTrades = userFills.filter(fill =>
            fill.oid.toString() === oid
        );

        if (orderTrades.length === 0) {
            return { totalPnl: 0, totalFees: 0, netPnl: 0, trades: [] };
        }

        let totalPnl = 0;
        let totalFees = 0;

        for (const trade of orderTrades) {
            const pnl = parseFloat(trade.closedPnl || '0');
            const fee = Math.abs(parseFloat(trade.fee || '0'));

            totalPnl += pnl;
            totalFees += fee;
        }

        const netPnl = totalPnl - totalFees;

        return {
            totalPnl,
            totalFees,
            netPnl,
            trades: orderTrades
        };

    } catch (error) {
        console.error('[getOrderPnl] Error fetching order PnL:', error);
        throw error;
    }
}
